USWDS-Site - WebP image conversion

amyleadem opened the pull request:
<!-- Please feel free to remove whatever sections/lines in this aren’t relevant.

Use the title line as the title of your pull request, then delete these lines. 

## Title line template: [Title]: Brief description

UI components: For pull requests that impact the look, feel, or functionality of USWDS itself, please open a pull request on the web-design-standards repo (https://github.com/uswds/uswds-site). 

-->

## Description
Closes: https://github.com/uswds/uswds-site/issues/1670

Preview link: [Home page](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.app.cloud.gov/preview/uswds/uswds-site/al-image-optimization/)

### Problem 
Using `.png` and `.jpg` image formats on our site results in unnecessarily large image file sizes, which slows down load times. This is reflected in low lighthouse scores on mobile.  Our current Lighthouse performance score on mobile is 68/100. 

Reference:
- [Lighthouse full report - Before](https://github.com/uswds/uswds-site/files/8919643/06-16-22-lighthouse--mobile.pdf)

### Solution
**Serve images in next-gen formats**
>Image formats like WebP and AVIF often provide better compression than PNG or JPEG, which means faster downloads and less data consumption.

By using next-gen image formats, we can optimize image compression and decrease load times. WebP was chosen because it has the broadest browser support. 

You can review the `Can I Use` browser-support reports for each considered image format:  [WebP](https://caniuse.com/webp), [AVIF](https://caniuse.com/avif), [JpegXL](https://caniuse.com/jpegxl)

#### Tasks
The `gulp-webp` dependency gives us a simple method to automate converting our existing `png` and `jpg` files to `webP`.

- [x] Install [gulp-webp ](https://www.npmjs.com/package/gulp-webp) as a devDependency to automate conversion to WebP
- [x] Add image conversion task to gulp
- [x] Convert png and jpg images to WebP format
- [x] Update image references on home page to WebP (Other images can be updated at a later time)
- [x] Resize images where possible to reduce file size
    - [x] Resized showcase images to from 1280 x 720 to 960 x 540 (Originally converted these to 640px but determined the loss of quality wasn't ideal)
- [x] Add height/width attributes to images to improve CLS
    - [More info on optimizing CLS](https://web.dev/optimize-cls/)

### Results
These updates resulted in an increased Lighthouse score with minimal impact on display quality. 

#### Lighthouse Ratings
**Before:**
Performance score: 68
[Lighthouse full report - Before](https://github.com/uswds/uswds-site/files/8919643/06-16-22-lighthouse--mobile.pdf)

**After**
Performance score: 89
[Lighthouse full report - After](https://github.com/uswds/uswds-site/files/8969427/Lighthouse-0623.pdf)

**Minimal loss of display quality**

![image](https://user-images.githubusercontent.com/93996430/175354150-5ac85248-915a-418b-889d-fc54d8cdcf97.png)

**Reduce image file size by more than half**

| Image              | Size before   | Size after |
|--------------------|--------|-------------------|
| feature-image.png  | 119KiB | 38KiB           |
| seedfund.jpg       | 185KiB | 73KiB             |
| plainlangauage.jpg | 81KiB  | 39KiB             |
| pubmed.jpg         | 82KiB  | 46KiB             |
| nasa.jpg           | 87KiB  | 53KiB             |
| va-healthcare.jpg  | 68KiB  | 27KiB             |
| vote.jpg           | 47KiB  | 17KiB             |
| **Total**           | **669KiB**  | **293KiB**             |

### Testing
To test, check that it has not had negative impact on image display in a variety of browsers in both desktop and mobile display. Run lighthouse to confirm that opportunities related to images haven't been missed. 

I also am interested in input on whether the following items have any potential downsides/disadvantages:
- Including the WebP conversion in our gulp build task
- Converting ALL files in the /img directory, not just the home page images

### Possible future tasks
If this method is approved, we should consider the following tasks:
- Update all `png` and `jpg` images on site to use WebP
- Update all `png` and `jpg` images in USWDS components to use WebP

## Additional information

Before you hit Submit, make sure you’ve done whichever of these applies to you:

- [ ] Follow the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `npm test` and make sure the tests for the files you have changed have passed.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.
- [x] Title your pull request using this format: [Website] - [UI component]: Brief statement describing what this pull request solves.

mejiaj commented on _includes/banner.html:
```diff
@@ -3,7 +3,7 @@
       <div class="usa-banner__inner">
         <div class="grid-col-auto">
-          <img class="usa-banner__header-flag" src="{{ site.baseurl }}/assets/img/us_flag_small.png" alt="U.S. flag">
+          <img class="usa-banner__header-flag" src="{{ site.baseurl }}/assets/img/us_flag_small.png" alt="U.S. flag" width="16" height="11">
```
Does this make a big difference for layout shift? If so, we should create a feature request in USWDS to add this to the component.

amyleadem replied:
It doesn't appear to have made a big difference in layout shift in this instance, but it does seem to be best practice (again). 

I have opened [this USWDS issue](https://github.com/uswds/uswds/issues/4831) so that we can return to the topic of adding height/width to our USWDS component image markup.

mejiaj replied:
Thanks for the clarification. I'm okay with including this since there's an issue to track it on USWDS.

mejiaj commented on _includes/footer.html:
```diff
@@ -93,7 +93,7 @@ <h3>Get support</h3>
           <div class="usa-identifier__logos">
               <a href="javascript:void(0);" class="usa-identifier__logo">
-                  <img class="usa-identifier__logo-img" src="{{ site.baseurl }}/img/gsa-logo.svg" alt="GSA logo" role="img">
+                  <img class="usa-identifier__logo-img" src="{{ site.baseurl }}/img/gsa-logo.svg" alt="GSA logo" role="img" width="48" height="48">
```
Same here, this is a core USWDS component. If it makes a difference we should make the same enhancement to the component.

mejiaj commented on package.json:
```diff
@@ -83,6 +83,7 @@
     "gulp-rename": "^2.0.0",
     "gulp-sourcemaps": "^3.0.0",
     "gulp-uglify": "^3.0.2",
+    "gulp-webp": "^4.0.1",
```
Last update was 01/23/19. We should keep a close eye on this in case dependabot alerts start popping up.

mejiaj reviewed (changes requested):
The images look good and I see we went from `2.8MB → 2.4MB` for page loads on homepage! 

Not sure if I'm testing correctly. I see a score of 68 for performance on the federalist preview and the fixes are outside of the scope of this issue.

Also, to prevent adding a bunch of new images to the repo, we should only commit the ones we're going to use on homepage (for now).

amyleadem commented on _config.yml:
```diff
@@ -8,7 +8,7 @@ federalist_base: "https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.app.cl
 federalist_component_preview: "iframe.html?id="
 federalist_release_prefix: "release-"
 meta:
-  og:image: /img/uswds-logo/lg-black.png
```
Thanks for the suggestion. It looks like `og:image` doesn't recognize SVG images (Reference: https://indieweb.org/The-Open-Graph-protocol#Does_not_support_SVG_images). This comment encouraged me to check if WebP is currently supported. 
For reference: 
- [Twitter explicitly supports WebP](https://developer.twitter.com/en/docs/twitter-for-websites/cards/overview/summary-card-with-large-image)
- Meta does not explicity support WebP in their docs, but does seem to support WebP in practice. From the [Meta sharing debugger](https://developers.facebook.com/tools/debug) tool:

![image](https://user-images.githubusercontent.com/93996430/175983546-bdb1d7a4-ad98-465a-965f-e7ef426412dd.png)

Because this does not have a direct impact on our performance at this time, I will revert this to the explicitly supported png file.

amyleadem commented on config/gulp/images.js:
```diff
@@ -15,4 +16,15 @@ function logImages(done) {
-exports.default = gulp.series(logImages, copyDocImages);
+function convertWebP() {
+  return gulp
+    .src('img/**/*.{png,jpg,jpeg}')
```
Removed `convertWebP` from the build task and have removed the unnecessary converted images from this PR. We can bring this task back into `build` in a separate issue when we are ready convert all relevant images to WebP.

mejiaj replied:
Great, thank you!

amyleadem commented:
Thanks for the comments, @mejiaj. I have pushed out the following updates for your review:
- Removed `convertWebP` from the gulp build task and instead exported for solo use. Updated glob to look only inside /home/showcase
- Removed the WebP images that don't appear on the home page. 
- Reverted `og:image` to use a `png` file.

>  I see a score of 68 for performance on the federalist preview and the fixes are outside of the scope of this issue.

In regards to lighthouse scores, I haven't seen a report as low as what you're seeing, but I am getting a wide range of scores for both federalist and the public site. Right now I see 88 in lighthouse for this branch. It does show a consistent score improvement in this branch over `develop`, and the published site seems to do better than federalist. Is the actual score a dealbreaker on this? If so, let me know and we can dig into this live.

mejiaj commented:
> 



> Thanks for the comments, @mejiaj. I have pushed out the following updates for your review:
> 
> * Removed `convertWebP` from the gulp build task and instead exported for solo use. Updated glob to look only inside /home/showcase
> * Removed the WebP images that don't appear on the home page.
> * Reverted `og:image` to use a `png` file.
> 
> > I see a score of 68 for performance on the federalist preview and the fixes are outside of the scope of this issue.
> 
> In regards to lighthouse scores, I haven't seen a report as low as what you're seeing, but I am getting a wide range of scores for both federalist and the public site. Right now I see 88 in lighthouse for this branch. It does show a consistent score improvement in this branch over `develop`, and the published site seems to do better than federalist. Is the actual score a dealbreaker on this? If so, let me know and we can dig into this live.

No not at all, I was just trying to understand where you were testing it (locally or federalist). These changes, regardless of score, are improvements.

mejiaj commented on pages/home.md:
```diff
@@ -68,15 +76,15 @@ graphic_list:
     <div class="grid-row grid-gap">
       <div class="tablet:grid-col">
-        <img src="{{ site.baseurl }}/img/home/showcase/pubmed.jpg" alt="">
+        <img src="{{ site.baseurl }}/img/webp/home/showcase/pubmed.webp" alt="Preview image of NASA Glenn Research Center home page" width="960" height="540">
```
Needs a unique description, currently shows the same as NASA above.

amyleadem replied:
Good catch! Updated in [7a40b5a](https://github.com/uswds/uswds-site/pull/1672/commits/7a40b5a8ea0083cc99d37f4eae8f5594fa27b1c4)

mejiaj reviewed (changes requested):
Thanks for these updates, just a minor note on a duplicate `alt` description on pubmed image.
